File: h2o/go/main.go
// Channel-based H2O implementation:
// Buffered channels limit hydrogen and oxygen admission. A channel barrier waits
// for exactly three admitted atoms before the tokens are released.
package main

import (
	"fmt"
	"sync"
)

type Barrier struct {
	arrive chan chan struct{}
}

func NewBarrier(n int) *Barrier {
	b := &Barrier{arrive: make(chan chan struct{})}
	go func() {
		waiters := make([]chan struct{}, 0, n)
		for ready := range b.arrive {
			waiters = append(waiters, ready)
			if len(waiters) == n {
				for _, waiter := range waiters {
					close(waiter)
				}
				waiters = make([]chan struct{}, 0, n)
			}
		}
	}()
	return b
}

func (b *Barrier) Wait() {
	ready := make(chan struct{})
	b.arrive <- ready
	<-ready
}

func main() {
	hydrogen := make(chan struct{}, 2)
	oxygen := make(chan struct{}, 1)
	barrier := NewBarrier(3)
	var wg sync.WaitGroup

	h := func() {
		defer wg.Done()
		hydrogen <- struct{}{}
		fmt.Println("H ready")
		barrier.Wait()
		<-hydrogen
	}
	o := func() {
		defer wg.Done()
		oxygen <- struct{}{}
		fmt.Println("O ready")
		barrier.Wait()
		<-oxygen
	}

	for i := 0; i < 4; i++ {
		wg.Add(1)
		go h()
	}
	for i := 0; i < 2; i++ {
		wg.Add(1)
		go o()
	}
	wg.Wait()
}

File: h2o/rust/main.rs
// Blocking H2O implementation:
// Counting semaphores built from Mutex and Condvar admit two hydrogens and one
// oxygen. A Barrier groups the three atoms before permits are returned.
use std::sync::{Arc, Barrier, Condvar, Mutex};
use std::thread;

struct CountingSemaphore {
    permits: Mutex<i32>,
    cond: Condvar,
}

impl CountingSemaphore {
    fn new(permits: i32) -> Self {
        Self {
            permits: Mutex::new(permits),
            cond: Condvar::new(),
        }
    }

    fn acquire(&self) {
        let mut permits = self.permits.lock().unwrap();
        while *permits == 0 {
            permits = self.cond.wait(permits).unwrap();
        }
        *permits -= 1;
    }

    fn release(&self) {
        let mut permits = self.permits.lock().unwrap();
        *permits += 1;
        self.cond.notify_all();
    }
}

fn main() {
    let hydrogen = Arc::new(CountingSemaphore::new(2));
    let oxygen = Arc::new(CountingSemaphore::new(1));
    let barrier = Arc::new(Barrier::new(3));
    let mut atoms = Vec::new();

    for _ in 0..4 {
        let hydrogen = Arc::clone(&hydrogen);
        let barrier = Arc::clone(&barrier);
        atoms.push(thread::spawn(move || {
            hydrogen.acquire();
            println!("H ready");
            barrier.wait();
            hydrogen.release();
        }));
    }
    for _ in 0..2 {
        let oxygen = Arc::clone(&oxygen);
        let barrier = Arc::clone(&barrier);
        atoms.push(thread::spawn(move || {
            oxygen.acquire();
            println!("O ready");
            barrier.wait();
            oxygen.release();
        }));
    }

    for atom in atoms {
        atom.join().unwrap();
    }
}
